components/resource_metering/src/reporter.rs

```diff
@@ -119,54 +125,84 @@ impl Runnable for ResourceMeteringReporter {
+                // Examples:
+                //
+                //   Case 1:
+                //   +---------------------+---------------------------------------+
+                //   | Max Resource Groups |   8                                   |
+                //   +----------+----------+-------+-------+-------+-------+-------+
+                //   |          |   Tag    | tag 3 | tag 2 | tag 5 | tag 4 | tag 1 |
+                //   | Records  +----------+-------+-------+-------+-------+-------+
+                //   |          | CPU Time |   90  |   60  |   50  |   30  |   10  |
+                //   +----------+----------+-------+-------+-------+-------+-------+
+                //   | Threshold           |   0                                   |
+                //   +----------+----------+-------+-------+-------+-------+-------+
+                //   | Will be Saved       |   Y   |   Y   |   Y   |   Y   |   Y   |
+                //   +---------------------+-------+-------+-------+-------+-------+
+                //
+                //  Case 2:
+                //   +---------------------+---------------------------------------+
+                //   | Max Resource Groups |   3                                   |
+                //   +----------+----------+-------+-------+-------+-------+-------+
+                //   |          |   Tag    | tag 3 | tag 2 | tag 5 | tag 4 | tag 1 |
+                //   | Records  +----------+-------+-------+-------+-------+-------+
+                //   |          | CPU Time |   90  |   60  |   50  |   30  |   10  |
+                //   +----------+----------+-------+-------+-------+-------+-------+
+                //   | Threshold           |   30                      ^           |
+                //   +----------+----------+-------+-------+-------+-------+-------+
+                //   | Will be Saved       |   Y   |   Y   |   Y   |   N   |   N   |
+                //   +---------------------+-------+-------+-------+-------+-------+
+                let threshold = (tmp_group_map.len() > max_resource_groups)
+                    .then(|| {
+                        tmp_top_vec.extend(tmp_group_map.values());
+                        let (_, threshold, _) = tmp_top_vec
+                            .select_nth_unstable_by(max_resource_groups, |a, b| b.cmp(a));
+                        *threshold
+                    })
+                    .unwrap_or(0);
 
-                if self.records.len() > self.config.max_resource_groups {
-                    self.find_top_k.clear();
-                    for (_, _, total) in self.records.values() {
-                        self.find_top_k.push(*total);
+                let timestamp_secs = records.begin_unix_time_secs;
+                let mut other = 0;
+                for (tag, ms) in tmp_group_map.drain() {
+                    if ms > threshold {
+                        let (ts, cpu) = self.records.entry(tag).or_insert((vec![], vec![]));
```

Comment:
Can we use `SmallVec` to reduce allocations?

Reply:
If so, it still needs to allocate later. 

https://github.com/tikv/tikv/blob/3044802f1171ca9edbd16f49214988e3a1a33b3c/components/resource_metering/src/reporter.rs#L241-L249